lib/src/front_end/ast_node_visitor.dart

```diff
@@ -357,7 +361,35 @@ class AstNodeVisitor extends ThrowingAstVisitor<void>
 
   @override
   void visitFormalParameterList(FormalParameterList node) {
-    throw UnimplementedError();
+    // Find the parameter immediately preceding the optional parameters (if
+    // there are any).
+    FormalParameter? lastRequired;
```

Comment:
[optional] Consider using the index instead

```dart
    var lastRequired =
        node.parameters.indexWhere((p) => p is DefaultFormalParameter) - 1;

    // ...

    for (var i = 0; i < node.parameters.length; i++) {
      builder.add(node.parameters[i]);

      // If the optional parameters start after this one, put the delimiter
      // at the end of its line.
      if (i == lastRequired) builder.leftDelimiter(node.leftDelimiter!);
    }
```

Reply:
Done, with some tweaks.